phoenix-core/src/it/java/org/apache/phoenix/execute/PartialCommitIT.java

```diff
@@ -0,0 +1,299 @@
+                                                    "select count(*) from " + TABLE_NAME_TO_FAIL + " where k = '" + ROW_TO_FAIL + "'",
+                                                    "select count(*) from " + TABLE_NAME_TO_FAIL + " where k = 'z'"), 
+                                       newArrayList(new Integer(4), new Integer(0), new Integer(1)));
+    }
+    
+    private void testPartialCommit(List<String> statements, int failureCount, int[] expectedUncommittedStatementIndexes, boolean willFail,
+                                   List<String> countStatementsForVerification, List<Integer> expectedCountsForVerification) {
+        Preconditions.checkArgument(countStatementsForVerification.size() == expectedCountsForVerification.size());
+        
+        try (Connection con = getConnectionWithTableOrderPreservingMutationState()) {
+            con.setAutoCommit(false);
+            Statement sta = con.createStatement();
+            for (String statement : statements) {
+                sta.execute(statement);
+            }
+            try {
+                con.commit();
+                if (willFail) {
+                    fail("Expected at least one statement in the list to fail");
+                } else {
+                    assertEquals(0, con.unwrap(PhoenixConnection.class).getStatementExecutionCounter()); // should have been reset to 0 in commit()
+                }
+            } catch (SQLException sqle) {
+                if (!willFail) {
+                    fail("Expected no statements to fail");
+                }
+                assertEquals(CommitException.class, sqle.getClass());
+                int[] uncommittedStatementIndexes = ((CommitException)sqle).getUncommittedStatementIndexes();
+                assertEquals(failureCount, uncommittedStatementIndexes.length);
+                assertArrayEquals(expectedUncommittedStatementIndexes, uncommittedStatementIndexes);
+            }
+            
+            // verify data in HBase
+            for (int i = 0; i < countStatementsForVerification.size(); i++) {
+                String countStatement = countStatementsForVerification.get(i);
+                ResultSet rs = sta.executeQuery(countStatement);
+                if (!rs.next()) {
+                    fail("Expected a single row from count query");
+                }
+                assertEquals(expectedCountsForVerification.get(i).intValue(), rs.getInt(1));
+            }
+        } catch (SQLException e) {
+            fail(e.toString());
+        }
+    }
+    
+    private PhoenixConnection getConnectionWithTableOrderPreservingMutationState() throws SQLException {
+        Connection con = driver.connect(url, new Properties());
+        PhoenixConnection phxCon = new PhoenixConnection(con.unwrap(PhoenixConnection.class));
```

Comment:
Thanks for the modifications, Eli. How about option (1), but have the existing constructors funnel through MutationState(int maxSize, PhoenixConnection connection) which would construct the HashMap while chaining through to your new constructor? Then you can make the Map final, but only initialize it in your new constructor? 

+1 after that if that sounds ok to you.